### PR TITLE
Feature/node bracket

### PR DESCRIPTION
**Pull request by staleyLANL**
Sorry, slamming y'all with another PR. See my long winded code comments for an explanation of what I'm doing. @whaeck may be able to use this material right away. Note that I haven't formally tested it yet, but I don't think it will brick your computer or anything.

### CODE REVIEW

**Comment by staleyLANL** [comment]
OK, I'm not sure what's going on here. It says 88 files changed. I added one and changed about 3. I may have made some sort of git mistake.

**Comment by staleyLANL** [comment]
Maybe the other PR that was approved earlier today didn't contain all the material I thought it did. I'd had two PRs in, and removed what I thought was the first, outdated one (because the second should have included the first's material). Anyway, the main new stuff is supposed to be a couple new classes near the bottom of node detail, and a new file node subscript.

**Comment by whaeck** [comment]
@staleyLANL I think the previous one did not get pulled into master but in feature/diagnostics-and-more.

**Comment by staleyLANL** [comment]
@whaeck Ah, so somehow the PR I merged earlier was slated to go into the other branch, not master. Then, since the current one was an extension of that - and apparently I did say to put it into master - then master thinks it's getting both the brand-new stuff and the stuff that only made it into diagnostics-and-more earlier? My main concern is that my work doesn't vanish somehow. :-) The new material should basically be the bottom part of detail.hpp, and everything in subscript.hpp, both in Node. See subscript.hpp's long comment for an explanation of the what/why/how. I think you can use this stuff right away.

**Comment by whaeck** [comment]
@staleyLANL To solve this, you may want to pull feature/diagnostics-and-more into master first, and then update this one to the new master once that is done.

File: src/GNDStk/Node/src/subscript.hpp
Comment (whaeck) [inline src/GNDStk/Node/src/subscript.hpp:177]: Now, if I wrote the following:
```
n[ axis, [] ( const auto& node ) { return node( index ) == "0"; } ]
```
would that return an lvalue so that I can actually modify that node?

Comment (whaeck) [reply src/GNDStk/Node/src/subscript.hpp:177]: I'm inclined to agree with you that if the operator[] returns more than one node, we should not treat the container as an lvalue. Having the vector contain reference_wrappers is a more sensible solution - unless we establish the rule that ONLY a single node can be an lvalue.

**Review by whaeck — APPROVED** [review]
Now that was a lot easier to review :-)

Do you have some tests associated to this? You may want to add it to the todo list for later, or add on a test file if you have time.

Other than that, I just have a question on using the operator[] with a filter to retrieve a single instance of a many node.

**Comment by staleyLANL** [comment]
> Now that was a lot easier to review :-)
> 
> Do you have some tests associated to this? You may want to add it to the todo list for later, or add on a test file if you have time.
> 
> Other than that, I just have a question on using the operator[] with a filter to retrieve a single instance of a many node.

I wrote some informal tests, but haven't integrated any tests into the formal test quite yet. That's of course on my to-do list; my intention had been to just get this in quickly so that it can be used.

File: src/GNDStk/Node/src/subscript.hpp
Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: I'm thinking more and more than the call operator should also return a vector of reference wrappers, not a vector of copies of the nodes.

And it goes without saying (although, well, I'm saying it) that that applies to the case where you use child_t<void>, not child_t<some type other than void>. In the latter case, conversion from node to some user type must be provided, and the conversion is applied to create each element of the output vector. So stuff is created on-the-fly, and the reference idea is not applicable.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: > Now, if I wrote the following:
> 
> ```
> n[ axis, [] ( const auto& node ) { return node( index ) == "0"; } ]
> ```
> 
> would that return an lvalue so that I can actually modify that node?

I think you want to *downgrade* axis to an allow::one, and *add* the lambda:
```
n[ --axis + lambda ];
```
Let me know if this indeed works. While I think I got `operator[]` right, I don't have full tests for it yet.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: Of course you'd presumably actually assign that to something:
```
n[ --axis + lambda ] = foobar;
```
Note that my `operator[]` is tricked out so that it can assign to either a value of the *type* given to (in this case) `axis`, or to anything that's implicitly convertible to that type (like, `int` to `double`) - or even to a string, if for whatever reason you'd like to put in the raw string directly instead of using the type you've associated with the keyword.

Comment (whaeck) [reply src/GNDStk/Node/src/subscript.hpp:177]: Ah, I forgot that I needed a '+' to add the filter. I'll give it a try when I update the v1.9 branch.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: Re: my above comment edit. If I ever make a comment that sounds stupid, let me know, and I can probably fix it. It's probably markdown's fault!!. Unless it's my fault, in which case I'll still try to fix it.

Comment (whaeck) [reply src/GNDStk/Node/src/subscript.hpp:177]: So if I write:
```
auto filter = [] ( const auto& node ) { return node( index ) == "0"; };
n[ Axis{} / --axis + filter ] = Axis( 0, "someLabel", "someUnit" );
```
I can use one of the higher level components to change the node itself (if I have a convert function handy, that is)?

**Comment by whaeck** [comment]
@staleyLANL I'll go and merge this tomorrow once I've done a few tests in the v1.9 branch.

File: src/GNDStk/Node/src/subscript.hpp
Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: By the way, regarding my `--` and `++` operators on keywords: Strangely enough, they *don't* actually modify the object itself, like how `--i` decrements to actual contents of the integer `i`. My keyword arithmetic always produces brand-new objects. (In many cases it couldn't *not* do that, because the arithmetic changes a template type!). So, `--child_t` has no effect on the original, and the original can even be `const`. Note that the language does indeed allow this behavior - of increment and decrement not changing the object to which they're applied. Of course I wouldn't do it if the language didn't allow it.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: I need to check if the remarks in my `foobar` example above actually work as I indent, re: the right-hand side being `std::string`. I may need to explicitly support the `char*` case, to prevent a general template function overload (from which an exact match can be produced) from being called, in place of the expected implicit `char *` to `std::string` conversion. Ah, technical stuff.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: To prevent any possible confusion... Be aware that a single minus (unary minus, `-`) in front of a keyword downgrades the *first* template argument - `type` - to `void`, meaning raw tree form rather than user-defined type. That applies to both `meta_t` and `child_t`. *Two* minus signs (predecrement, `--`) in front of a keyword downgrades the *second* template argument of a `child_t` - making it `allow::one`, regardless of its previous value. (The predecrement doesn't apply to `meta_t` objects, which don't have the `one` versus `many` distinction.)

There's indeed a pattern to all this funny stuff I wrote. A user just needs to stare at it for a bit, in order for it all to make sense. If it still doesn't make sense after some study, then let's talk about it. :-)

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: > So if I write:
> 
> ```
> auto filter = [] ( const auto& node ) { return node( index ) == "0"; };
> n[ Axis{} / --axis + filter ] = Axis( 0, "someLabel", "someUnit" );
> ```
> 
> I can use one of the higher level components to change the node itself (if I have a convert function handy, that is)?

Offhand that looks correct, and it *should* work. Let me know if it doesn't! The operator business is fairly recent, and the bracket operator is brand-new, so there could always be a glitch here or there.

Of course we can, as I illustrated in my sample high-level Axis, associate an existing node with a high-level structure, and have a `save()` or other operation that changes the node automatically. We'll probably want to allow more than one way to do things.

Your comment reminds me: we should have a discussion sometime of what I did provide in the way of built-in `convert` functions. Roughly speaking: a handful of basic ones are provided for metadata (string/type conversions), and none are provided for children (node/type) conversions. None, because I can't imagine any meaningful default for that case; it depends on what's in somebody's data structure.

Comment (staleyLANL) [reply src/GNDStk/Node/src/subscript.hpp:177]: It looks like all this operator business has proven to be even more useful than I anticipated it might be when I wrote it. That's a good thing.

**Comment by staleyLANL** [comment]
@whaeck @jlconlin Looks like this has one approval, at least, but I still don't have merge permissions. What do you think about going ahead and merging? I'll have additional material soon, and probably more tests for the above, by week's end, at which time I can put in another PR.

**Pull request merged**
